### PR TITLE
Bug 1813198: Windows CDs use VirtIO instead of sata

### PR DESCRIPTION
**Pull request by glekner**
This should later be fixed when `common-templates` have validations for CDs

### CODE REVIEW

**Comment by glekner** [comment]
@suomiy @yaacov can you review?

**Comment by glekner** [comment]
this should fix warning messages - https://github.com/kubevirt/common-templates/pull/148

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/virtual-hardware-tab/virtual-hardware-tab.tsx [hunk 1]
```diff
@@ -86,7 +94,7 @@ const VirtualHardwareTabFirehose: React.FC<VirtualHardwareTabFirehoseProps> = ({
   const diskWrapper = DiskWrapper.initializeFromSimpleData({
     name: availableCDName,
     type: DiskType.CDROM,
-    bus: templateValidations.getDefaultBus(),
+    bus: os?.startsWith('win') ? DiskBus.SATA : templateValidations.getDefaultBus(),
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/virtual-hardware-tab/virtual-hardware-tab.tsx:97]: so this will be later deprecated in favor of common templates right? Can you please write a TODO comment somewhere with the PR link: https://github.com/kubevirt/common-templates/pull/148. Together with bugzilla bug to track this.

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/virtual-hardware-tab/virtual-hardware-tab.tsx [hunk 2]
```diff
@@ -185,6 +194,7 @@ type VirtualHardwareConnectedProps = VirtualHardwareTabFirehoseProps & {
 
 const stateToProps = (state, { wizardReduxID }) => ({
   namespace: iGetCommonData(state, wizardReduxID, VMWizardProps.activeNamespace),
+  os: iGetVmSettingAttribute(state, wizardReduxID, VMSettingsField.OPERATING_SYSTEM),
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/virtual-hardware-tab/virtual-hardware-tab.tsx:197]: please use `iGetVmSettingValue` instead

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/virtual-hardware-tab/virtual-hardware-tab.tsx (hunk 1 above)
Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/virtual-hardware-tab/virtual-hardware-tab.tsx:97]: can we do toLowerCase and includes instead?

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx [hunk 3]
```diff
@@ -581,6 +587,7 @@ export type DiskModalProps = {
   showInitialValidation?: boolean;
   isCreateTemplate?: boolean;
   isEditing?: boolean;
+  isWindows?: boolean;
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx:590]: please, let's not add more properties to this component. We can pass the desired value inside the disk value in place where we need it.

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx [hunk 4]
```diff
@@ -458,7 +464,7 @@ export const DiskModal = withHandlePromise((props: DiskModalProps) => {
               isValid={!isValidationError(busValidation)}
               value={asFormSelectValue(bus)}
               id={asId('interface')}
-              isDisabled={isDisabled('interface')}
+              isDisabled={isDisabled('interface') || isWindowsCD}
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx:467]: IMO we should still allow the user to change it. Anyway generally, it is better to pass these values in editConfig instead from outside

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx [hunk 5]
```diff
@@ -458,7 +464,7 @@ export const DiskModal = withHandlePromise((props: DiskModalProps) => {
               isValid={!isValidationError(busValidation)}
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx:464]: btw shouldn't we change TemplateValidations to react to cd-roms in this PR as well. And allow sata's for now until common-templates are fixed?

File: frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx (hunk 4 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx:467]: Why give the illusion that the user can change the interface?
the only wanted bus for windows && cd is SATA so it should be fixed to SATA.
thats why this change shouldn't be deprecated, IMO

Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/components/modals/disk-modal/disk-modal.tsx:467]: what about scsi? IMo we should use only what comes from the template and don't have a special logic like this one

**Comment by glekner** [comment]
@suomiy I'm trying to make this work on my local env with the updated templates.

I didn't want to pass a disk to the TemplateValidation constructor as it seems out of place, I ended up cloning methods for CD type. This works for creating a VM from scratch, but when choosing a Windows Template I created, I still see the unsupported buses. can you take a look?

Thanks!

**Comment by suomiy** [comment]
> I didn't want to pass a disk to the TemplateValidation constructor as it seems out of place, I ended up cloning methods for CD type. This works for creating a VM from scratch, but when choosing a Windows Template I created, I still see the unsupported buses. can you take a look?

This happens because the created template doesn't include the validations anymore. We should probably copy the validations when we are creating the template from common templates.

File: frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts [hunk 6]
```diff
@@ -14,7 +14,12 @@ export class ValidationJSONPath extends ObjectEnum<string> {
   static readonly MEMORY = new ValidationJSONPath(
     'jsonpath::.spec.domain.resources.requests.memory',
   );
-  static readonly BUS = new ValidationJSONPath('jsonpath::.spec.domain.devices.disks[*].disk.bus');
+  static readonly DISKBUS = new ValidationJSONPath(
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts:17]:
```suggestion
  static readonly DISK_BUS = new ValidationJSONPath(
```

\+ CD_BUS

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts [hunk 7]
```diff
@@ -97,10 +97,14 @@ export const internalStorageDiskBusUpdater = ({
...
-      const resultValidation = newValidations.validateBus(new DiskWrapper(disk).getDiskBus());
+      const diskWrapper = new DiskWrapper(disk, true);
+      const resultValidation = newValidations.validateBus(
+        diskWrapper.getType(),
+        diskWrapper.getDiskBus(),
+      );
       if (!resultValidation.isValid && resultValidation.type === ValidationErrorType.Error) {
         someBusChanged = true;
-        finalDisk = new DiskWrapper(disk, true)
+        finalDisk = diskWrapper
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts:107]: can you please revert this change? This is done, so there is no unnecessary copying when the validation passes fine

File: frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts [hunk 8]
```diff
@@ -47,6 +51,17 @@ export class TemplateValidations {
     return new Set(allowedBuses.length === 0 ? DiskBus.getAll() : allowedBuses);
   };
 
+  getAllowedCDBuses = (
```
[the remark this answers is not in the document]
Comment (suomiy) [reply frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts:56]: still relevant, can we just pass the DiskType and not have the same logic twice?

File: frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts [hunk 9]
```diff
@@ -66,36 +91,47 @@ export class TemplateValidations {
...
     const allowedBuses = this.getAllowedBuses();
+    const allowedCDBuses = this.getAllowedCDBuses();
     const recommendedBuses = this.getRecommendedBuses();
+    const recommendedCDBuses = this.getRecommendedCDBuses();
     const otherAllowedBuses = otherTempValidations.getAllowedBuses();
+    const otherAllowedCDBuses = otherTempValidations.getAllowedCDBuses();
     const otherRecommendedBuses = otherTempValidations.getRecommendedBuses();
+    const otherRecommendedCDBuses = otherTempValidations.getRecommendedCDBuses();
 
     return (
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts:102]: maybe we could do a lazy compare here? First get AllowedBuses and compare if not equal return. Then get aand compare recommended. Etc.

File: frontend/report.20200518.185441.7687.0.001.json [hunk 10]
```diff
@@ -0,0 +1,764 @@
+
```
Comment (suomiy) [inline frontend/report.20200518.185441.7687.0.001.json:1]: runaway file?

**Comment by glekner** [comment]
- applied suggestions and added validations to templates
@suomiy

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts [hunk 11]
```diff
@@ -97,11 +97,16 @@ export const internalStorageDiskBusUpdater = ({
         VMWizardStorageType.WINDOWS_GUEST_TOOLS,
       ].includes(type)
     ) {
-      const resultValidation = newValidations.validateBus(new DiskWrapper(disk).getDiskBus());
+      const resultValidation = newValidations.validateBus(
+        new DiskWrapper(disk).getType(),
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts:101]: you can make a variable for these two lines

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts [hunk 12]
```diff
@@ -97,11 +97,16 @@ export const internalStorageDiskBusUpdater = ({
...
+        new DiskWrapper(disk).getDiskBus(),
+      );
       if (!resultValidation.isValid && resultValidation.type === ValidationErrorType.Error) {
         someBusChanged = true;
         finalDisk = new DiskWrapper(disk, true)
-          .appendTypeData({ bus: newValidations.getDefaultBus().getValue() })
+          .appendTypeData({
+            bus: newValidations.getDefaultBus(new DiskWrapper(disk).getType()).getValue(),
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts:108]: and for getting the type here

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts [hunk 13]
```diff
@@ -97,11 +97,16 @@ export const internalStorageDiskBusUpdater = ({
...
       ].includes(type)
     ) {
-      const resultValidation = newValidations.validateBus(new DiskWrapper(disk).getDiskBus());
+      const resultValidation = newValidations.validateBus(
+        new DiskWrapper(disk).getType(),
+        new DiskWrapper(disk).getDiskBus(),
+      );
       if (!resultValidation.isValid && resultValidation.type === ValidationErrorType.Error) {
         someBusChanged = true;
         finalDisk = new DiskWrapper(disk, true)
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts:106]: the main difference is that the final disk is mutated, so it needs to be copied to its own wrapper

File: frontend/packages/kubevirt-plugin/src/constants/vm/constants.ts [hunk 14]
```diff
@@ -5,6 +5,7 @@ export const ANNOTATION_FIRST_BOOT = 'kubevirt.ui/firstBoot';
 export const ANNOTATION_DESCRIPTION = 'description';
 export const ANNOTATION_PXE_INTERFACE = 'kubevirt.ui/pxeInterface';
 export const CUSTOM_FLAVOR = 'Custom';
+export const VALIDATIONS = 'validations';
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/constants/vm/constants.ts:8]: +1 for the constant. Can we add `ANNOTATION` to its name?

File: frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts [hunk 15]
```diff
@@ -50,6 +51,7 @@ export const initializeCommonMetadata = (
   if (template) {
     entity.addLabel(LABEL_USED_TEMPLATE_NAME, getName(template));
     entity.addLabel(LABEL_USED_TEMPLATE_NAMESPACE, getNamespace(template));
+    entity.addAnotation(VALIDATIONS, getAnnotations(template)?.[VALIDATIONS]);
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts:54]: can you please make a new function called initializeCommonTemplateMetadata and put it there instead? 
This function is also used for creating VMs and we are not really expecting VMs to include validations.

File: frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts [hunk 16]
```diff
@@ -63,50 +70,67 @@ export class TemplateValidations {
     if (this === otherTempValidations) {
       return true;
     }
+    const isBusesEqual = (buses, otherBuses) =>
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts:73]: +1 for having a function. Can we move it one layer up to the class and make it private?

File: frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts [hunk 17]
```diff
@@ -63,50 +70,67 @@ export class TemplateValidations {
...
-    const otherAllowedBuses = otherTempValidations.getAllowedBuses();
-    const otherRecommendedBuses = otherTempValidations.getRecommendedBuses();
-
-    return (
-      allowedBuses.size === otherAllowedBuses.size &&
-      recommendedBuses.size === otherRecommendedBuses.size &&
-      [...allowedBuses].every((bus) => otherAllowedBuses.has(bus)) &&
-      [...recommendedBuses].every((bus) => otherRecommendedBuses.has(bus))
-    );
+    const allowedBuses = this.getAllowedBuses(DiskType.DISK);
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/utils/validations/template/template-validations.ts:77]: We could also do `[DiskType.DISK, DiskType.CDROM].every` but I do not mind either way

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts [hunk 18]
```diff
@@ -97,11 +97,15 @@ export const internalStorageDiskBusUpdater = ({
         VMWizardStorageType.WINDOWS_GUEST_TOOLS,
       ].includes(type)
     ) {
-      const resultValidation = newValidations.validateBus(new DiskWrapper(disk).getDiskBus());
+      const diskType = new DiskWrapper(disk).getType();
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/redux/state-update/storage-tab-state-update.ts:100]: better, but still the diskWrapper can be reused

File: frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts [hunk 19]
```diff
@@ -91,3 +92,18 @@ export const initializeCommonVMMetadata = (
...
+    [VMSettingsField.DESCRIPTION]: string;
+    [VMSettingsField.FLAVOR]: string;
+    [VMSettingsField.WORKLOAD_PROFILE]: string;
+    osID: string;
+    osName: string;
+  },
+  entity: VMWrapper | VMTemplateWrapper,
+  template?: TemplateKind,
+) => {
+  initializeCommonMetadata(settings, entity, template);
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts:107]: please use the same pattern as VM and call the `initializeCommonMetadata` separately

File: frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts [hunk 20]
```diff
@@ -91,3 +92,18 @@ export const initializeCommonVMMetadata = (
     );
   }
 };
+
+export const initializeCommonTemplateMetadata = (
+  settings: {
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts:97]: in that case you can remove the settings here and it will simplify the API

File: frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts [hunk 21]
```diff
@@ -91,3 +92,18 @@ export const initializeCommonVMMetadata = (
     );
   }
 };
+
+export const initializeCommonTemplateMetadata = (
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts:96]: this function should be called - to initialize common template metadata

File: frontend/packages/kubevirt-plugin/src/utils/validations/template/utils.ts [hunk 22]
```diff
@@ -67,3 +67,6 @@ export const combineIntegerValidationResults = (
   }
   return asValidationObject(message, finalType);
 };
+
+export const isSetEqual = (set: Set<any>, otherSet: Set<any>) =>
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/utils/validations/template/utils.ts:71]: this can be rather moved to `src/utils/common.ts`

File: frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts [hunk 23]
```diff
@@ -91,3 +94,11 @@ export const initializeCommonVMMetadata = (
...
+  entity: VMWrapper | VMTemplateWrapper,
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts:99]:
```suggestion
  entity: VMTemplateWrapper,
```

File: frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts [hunk 24]
```diff
@@ -91,3 +94,11 @@ export const initializeCommonVMMetadata = (
...
+  template?: TemplateKind,
+) => {
+  entity.addLabel(TEMPLATE_TYPE_LABEL, TEMPLATE_TYPE_VM);
+  entity.addAnotation(ANNOTATION_VALIDATIONS, getAnnotations(template)?.[ANNOTATION_VALIDATIONS]);
```
Comment (suomiy) [inline frontend/packages/kubevirt-plugin/src/k8s/requests/vm/create/common.ts:103]: we should check for undefined/null here so we don't add epmty annotations label

**Comment by yaacov** [comment]
@glekner hi, ^^ looks like this should be backported by hand :-(
